[Jovyn Tan] iP

jovyntls opened the pull request:
# Duke Chatbot :+1: :sparkles: :camel: :tada: :rocket: :metal: :octocat: 😊

According to the [nus-cs2103-AY2122S1/ip README.md](https://github.com/nus-cs2103-AY2122S1/ip/blob/master/README.md),

> This is a project template for a greenfield Java project. It's named after the Java mascot Duke. 

## Usage

This **chatbot** can be _used by_: 
* the JAR file, found [here](https://github.com/jovyntls/ip/blob/master/build/libs/ip-1.0-SNAPSHOT-all.jar)
* run tests using `bash ./text-ui-test/duke-test.sh`
* clone the repo and run the `main` function:

```java
public static void main(String[] args) {
    String logo = " ____        _        \n"
            + "|  _ \\ _   _| | _____ \n"
            + "| | | | | | | |/ / _ \\\n"
            + "| |_| | |_| |   <  __/\n"
            + "|____/ \\__,_|_|\\_\\___|\n";
    System.out.println("Hello from\n" + logo);

    Duke duke = new Duke("../../data/duke_storage.txt");

    duke.greet();
    duke.taskMode();
}
```

## Progress

### Milestones
1. ~~Week 1~~
2. ~~Week 2~~
3. ~~Week 3~~
4. Week 4

### To do

- [x] Basic chatbot implementation
- [x] Extra features
- [ ] GUI

jovyntls committed:
Ignore out/ directory

jovyntls committed:
Add greeting to the user

jovyntls committed:
Add ChatBot class

jovyntls committed:
Add echoing and exiting

jovyntls committed:
Refactor Duke class to use instance methods

jovyntls committed:
Add Task and TaskList classes

jovyntls committed:
Add ability to create tasks

jovyntls committed:
Add ability to list tasks

jovyntls committed:
Fix javadocs for toString method in TaskList

jovyntls committed:
Add ability to complete tasks

jovyntls committed:
Fix wording and spacing for messages

jovyntls committed:
Add ToDos, Events and Deadlines

jovyntls committed:
Fix bug in task titles for events and deadlines

jovyntls committed:
Improve spacing and readability for output messages

jovyntls committed:
Add seed data for automated tests

jovyntls committed:
Add error handling and DukeException

jovyntls committed:
Trim whitespace for task inputs

jovyntls committed:
Refactor to use Collections ArrayList

jovyntls committed:
Add delete functionality

jovyntls committed:
Improve input handling

jovyntls committed:
Remove unnecessary throwing of DukeException

jovyntls committed:
Refactor input handling logic to be cleaner

jovyntls committed:
Add comments for each block

jovyntls committed:
Fix output message for deleting tasks

jovyntls committed:
Add test case for delete

jovyntls committed:
Add enums for Task types

jovyntls committed:
Refactor enums

jovyntls committed:
Add getIntInputAfterCommand to neaten task handling logic

jovyntls committed:
Fix linting errors from checkstyle

jovyntls committed:
Change Chatbot class to ChatbotUI interface

jovyntls committed:
Add methods for handling file data

jovyntls committed:
Refactor Task subclasses constructors

jovyntls committed:
Add overloaded constructors for task subclasses

jovyntls committed:
Rename methods to be more descriptive

jovyntls committed:
Add methods for handling data

jovyntls committed:
Handle loading and saving of data for Tasks

jovyntls committed:
Fix test cases

jovyntls committed:
Add test cases for loading saved tasks

jovyntls committed:
Reorganise test files

Currently, the test files are all in the same directory with a flat
structure.

This is disorganised and hard to read.

This commit will organise test files into subdirectories.

Reorganisation makes it easier to navigate the test files.

The test should be run from the project root directory.
The file to run is `./text-ui-test/duke-test.sh`

jovyntls committed:
Allow Duke to understand dates and times

jovyntls committed:
Merge branch 'branch-Level-7'

jovyntls committed:
Fix merge conflicts in Task subclasses

jovyntls committed:
Fix dates in test cases

jovyntls committed:
Separate Date and String parsing

jovyntls committed:
Add Parser class and lint

jovyntls committed:
Add Storage class

jovyntls committed:
Reorganise classes into packages

jovyntls committed:
Fix tests to work with files in packages

jovyntls committed:
Add Javadocs and fix checkstyle issues

jovyntls committed:
Remove .class files

jovyntls committed:
Add JUnit tests for Deadlines and Events

jovyntls committed:
Add JUnit tests for Parser interface

jovyntls committed:
Add packaged app as JAR file

jovyntls committed:
Add JavaDocs for public classes

jovyntls committed:
Refactor to follow CS2103 coding standard

jovyntls committed:
Add filtering for tasks by keyword

jovyntls committed:
Add filtering to test cases

jovyntls committed:
Merge branch 'branch-A-JavaDoc'

jovyntls committed:
Merge branch 'branch-A-CodingStandard'

jovyntls committed:
Merge branch 'branch-Level-9'

jovyntls committed:
Update README.md

jovyntls committed:
Add checkstyle command to Gradle

jovyntls committed:
Merge branch 'add-checkstyle-to-gradle'

jovyntls committed:
Edit README.md

jovyntls committed:
Add basic functioning GUI

jovyntls committed:
Add GUI for echo mode

jovyntls committed:
Implement basic task management with GUI

jovyntls committed:
Fix path for data storage

jovyntls committed:
Reorganise logic for Duke GUI

jovyntls committed:
Merge branch 'branch-Level-10'

jovyntls committed:
Add varargs to printMessage

jovyntls committed:
Merge branch 'branch-varargs'

yeo-yiheng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,174 @@
+import duke.gui.DukeDialogBox;
+import duke.gui.DukeGui;
+import duke.gui.UserDialogBox;
+import duke.task.Task;
```
Perhaps you could consider grouping the imports of a similar package together and separating the rest?

yeo-yiheng commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+     */
+    public String addNewDeadline(String taskTitle) {
+        int delimiter = taskTitle.indexOf("/by ");
+        LocalDate due = LocalDate.parse(taskTitle.substring(delimiter + 4));
```
Perhaps you could define a final constant that takes the value of 4 so we know what the value is referring to?

KT27Learn replied:
Perhaps you could add this 4 as a final constant to improved readability? I've noticed this in other areas as well.

yeo-yiheng commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+     */
+    public String addNewEvent(String taskTitle) {
+        int delimiter = taskTitle.indexOf("/at ");
+        LocalDate due = LocalDate.parse(taskTitle.substring(delimiter + 4));
```
Perhaps you could define a final constant to store the value of 4? I noticed this in other areas as well.

yeo-yiheng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,42 @@
+     * @throws DukeException if that the description is invalid
+     */
+    static String getStringFrom(String command, String message) throws DukeException {
+        if (message.length() <= command.length()) {
```
Maybe you could store the condition into a variable stating what it is checking for?

yeo-yiheng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,42 @@
+            );
+        }
+        String description = message.substring(command.length()).trim();
+        if (description.length() == 0) {
```
Maybe you could name this condition by storing it into a variable? I have noticed this in other areas as well.

yeo-yiheng reviewed (commented):
LGTM. I like that your program is really organized and clean. Just a little bit of nit-picking.

jovyntls committed:
Remove unused packages directory

jovyntls committed:
Fix gradle

jiayi1129 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,74 @@
+    private final String dataStoragePath;
+
+    /**
+     * A constructor for Duke Storage.
+     * @param dataStoragePath the path of the file where Duke data is stored.
```
Perhaps can consider adding a line between the description and parameter section?

jiayi1129 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    /**
+     * Writes a given String to a given file, without overwriting existing content.
+     * @param content The content to be written to the file
```
Perhaps can consider adding a line between the description and parameter section? I noticed the same issue in several other places too.

jiayi1129 commented on src/test/java/DeadlineTest.java:
```diff
@@ -0,0 +1,21 @@
+
+public class DeadlineTest {
+    @Test
+    public void createNewDeadlineTest(){
```
Perhaps you can consider naming your tests in the following format `featureUnderTest_testScenario_expectedBehavior()`? I noticed the same issue in several other places too.

jiayi1129 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,42 @@
+     * @return an int representing the description of the command. (e.g. "2")
+     * @throws DukeException if the description is invalid
+     */
+    static int getIntFrom(String command, String message) throws DukeException {
```
Perhaps the method name can be more specific such as `getIntFromCommand`? Let me know your thoughts on this.

jiayi1129 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    /**
+     * Parses a user-input for a number supplied after a given command.
+     * @param command A String command that comes before the desired number. (e.g. "delete")
+     * @param message The String that is the full input by the user. (e.g. "delete 2")
+     * @return an int representing the description of the command. (e.g. "2")
+     * @throws DukeException if the description is invalid
```
I like how clear this comment is! It really helps me in understanding what the method does and accomplishes easily.

jiayi1129 reviewed (commented):
Overall I like your program! Its it extremely well documented and thus easy to understand! Most of the code is intuitive and and its easy to understand the flow and logic too!

Other than a few minor nitpicks, LGTM! 😄

KT27Learn commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,74 @@
+    private final String dataStoragePath;
+
+    /**
+     * A constructor for Duke Storage.
```
Perhaps you could consider adding a line to separate the description and parameter in the javadocs? I noticed this in other areas as well.

KT27Learn commented on src/main/java/duke/gui/DukeDialogBox.java:
```diff
@@ -0,0 +1,24 @@
+import javafx.scene.control.Label;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+
```
Perhaps you could add a header comment to describe this class? I've noticed this in other areas as well.

KT27Learn commented on src/main/java/duke/gui/UserDialogBox.java:
```diff
@@ -0,0 +1,24 @@
+    private Label text;
+    private ImageView displayPicture;
+
+    public UserDialogBox(Label l, ImageView iv) {
```
Perhaps you could add in a header comment for this constructor?

KT27Learn commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,140 @@
+    protected char typeIndicator;
+    protected LocalDate timeDue;
+
+    protected enum TypeIndicator {
```
I like the use of Enum to indicate the type of task!

KT27Learn commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,140 @@
+        EVENT('E'),
+        DEADLINE('D');
+
+        private final char indicator;
```
Perhaps you could change the final variable to all caps?

KT27Learn reviewed (commented):
Everything looks great, and I think that it is an overall job well done! There are no major issues, and the code base was easy to understand. There are just some small style pointers that was brought up.

jovyntls committed:
Add JAR file

jovyntls committed:
Refactor DukeGui to improve readability

Currently, the handleUserInput method in DukeGui is easily confused with
the handleUserInput method in Duke (which extends DukeGui). Neither
overrides the other and they both accomplish different functionalities.

The handleUserInput method in DukeGui needs to be renamed and refactored
to show that it simply echoes the user input, and does not carry out any
additional computation about it.

Let's rename handleUserInput in DukeGui to 'echo', and remove the
dukeText variable to show that the same text is used for input and
output.

This renaming and refactoring clearly shows the difference between the
logic in DukeGui and Duke. Confusion is reduced since Duke extends
DukeGui and we clearly show that there should be no method overriding in
handleUserInput.

jovyntls committed:
Add assertions for Parser

jovyntls committed:
Reorganise GUI classes

jovyntls committed:
Remove deadcode from deprecated text UI

jovyntls committed:
Remove unnecessary 'this' from Duke class

jovyntls committed:
Refactor task classes

jovyntls committed:
Remove unnecessary this from TaskList class

jovyntls committed:
Fix errors from checkstyle

jovyntls committed:
Use Assertions

Add assertions for Parser

jovyntls committed:
Merge pull request #3 from jovyntls/branch-A-CodeQuality

Refactor to improve code quality

jovyntls committed:
Disallow adding of duplicate tasks

Duplicate tasks can be added to the task list without any warning or
exceptions raised by Duke.

Duplicated tasks may cause confusion for the user. Furthermore, if
features such as sorting (unstable) are introduced in the future, the
duplicated tasks may cause bugs that are hard to fix.

Raising an exception when the user attempts to add a duplicate task
reminds the user that the task is duplicated and prevents user errors.
Disallowing duplicated tasks also reduces the possibility of bugs in the
future.

Let's raise a user-facing DukeException when the user attempts to add a
task that has already been added, which allows the user to know that
duplicated tasks are not allowed.

Raising a DukeException is a clean way to handle the duplicated task.
Since it is user-facing, the user is also informed that such behaviour
is not allowed.

Additional information: Streams are used to check for duplicates.

jovyntls committed:
improve GUI chat text fields

jovyntls committed:
Add javadocs for DialogBoxStyle

jovyntls committed:
Add UI screenshot

jovyntls committed:
Add User Guide

jovyntls committed:
Fix path for data storage

jovyntls committed:
Add newest release of Duke

jovyntls committed:
Update README.md
